### PR TITLE
Behold uvbloop_t, a cross-platform event loop

### PR DESCRIPTION
**Pull request by rossdylan**
- Created basic event loop interface, uvbloop_t
- Implemented uvbloop with kqueue and epoll
- Refactored timer subsystem to use uvbloop
- Added 'rd_' prefix to linked list implementation
    * Most of the list macros conflict with things in OSX and FreeBSD
- Swapped out epoll for uvbloop in server code

### CODE REVIEW

**Comment by rossdylan** [comment]
This PR addresses issue #38 We support kqueue on OSX and FreeBSD and EPoll on Linux.

File: include/uvbloop.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * Unregister a timer from this eventloop. This is done by passing in an event
+ */
+int uvbloop_unregister_timer(uvbloop_t *loop, int id);
+
+/**
+ * Wait for events from the given uvbloop_t
+ */
+int uvbloop_wait(uvbloop_t *loop, uvbloop_event_t *events, uint64_t max_events);
```
Comment (robgssp) [inline include/uvbloop.h]: return value and `max_events` should be the same type. kevent and epoll are both int, so `max_events` should probably also be.

File: src/epoll_uvbloop.c [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+
+bool uvbloop_event_error(uvbloop_event_t *event) {
+    struct epoll_event e = *(struct epoll_event *)event;
+    return e.events & EPOLLERR || e.events & EPOLLHUP || !(e.events & EPOLLIN);
+}
+
+
+void *uvbloop_event_data(uvbloop_event_t *event) {
+    return ((struct epoll_event *)event)->data.ptr;
```
Comment (robgssp) [inline src/epoll_uvbloop.c]: why is this cast necessary?

File: src/timers.c [hunk 3]
```diff
@@ -1,32 +1,30 @@
...
     timer_entry_t *entry = NULL;
-    uint64_t exp;
     while(true) {
-        waiting = epoll_wait(p->epoll_fd, events, MAXEVENTS, -1);
+        waiting = uvbloop_wait(p->loop, events, MAXEVENTS);
         if(waiting > 0) {
             for(int i=0; i<waiting; i++) {
-                entry = (timer_entry_t *)events[i].data.ptr;
-                read(entry->tfd, &exp, sizeof(uint64_t));
+                entry = (timer_entry_t *)uvbloop_event_data(events);
```
Comment (robgssp) [inline src/timers.c]: should this be `events+i`?

Comment (robgssp) [inline src/timers.c]: Or rather, this being the timer thread, uvbloop_wait nevents could probably just be 1

Comment (rossdylan) [inline src/timers.c]: yep, good catch, I've added the [i] back in

File: src/epoll_uvbloop.c (hunk 2 above)
Comment (rossdylan) [inline src/epoll_uvbloop.c]: it was not, I think it was left over from when I was futzing around with the typedefs

**Pull request merged**